packages/storage-service/src/StorageService.ts

```diff
@@ -0,0 +1,217 @@
+  readonly #storage: StorageAdapter;
+
+  /**
+   * Constructs a new StorageService.
+   *
+   * @param options - The options.
+   * @param options.messenger - The messenger suited for this service.
+   * @param options.storage - Storage adapter for persisting data.
+   * If not provided, uses InMemoryStorageAdapter (data lost on restart).
+   */
+  constructor({ messenger, storage }: StorageServiceOptions) {
+    this.name = SERVICE_NAME;
+    this.#messenger = messenger;
+    this.#storage = storage ?? new InMemoryStorageAdapter();
+
+    // Warn if using in-memory storage (data won't persist)
+    if (!storage) {
+      console.warn(
+        `${SERVICE_NAME}: No storage adapter provided. Using in-memory storage. ` +
+          'Data will be lost on restart. Provide a storage adapter for persistence.',
+      );
+    }
+
+    // Register messenger actions
+    this.#messenger.registerMethodActionHandlers(
+      this,
+      MESSENGER_EXPOSED_METHODS,
+    );
+  }
+
+  /**
+   * Store large data in storage.
+   *
+   * ⚠️ **Designed for large values (100KB+), not many small ones.**
+   * Each storage operation has I/O overhead. For best performance,
+   * store one large object rather than many small key-value pairs.
+   *
+   * @example Good: Store entire cache as one value
+   * ```typescript
+   * await service.setItem('TokenList', 'cache', { '0x1': [...], '0x38': [...] });
+   * ```
+   *
+   * @example Avoid: Many small values
+   * ```typescript
+   * // ❌ Don't do this - too many small writes
+   * await service.setItem('TokenList', 'cache:0x1', [...]);
+   * await service.setItem('TokenList', 'cache:0x38', [...]);
+   * ```
+   *
+   * @param namespace - Controller namespace (e.g., 'SnapController').
+   * @param key - Storage key (e.g., 'npm:@metamask/example-snap:sourceCode').
+   * @param value - Data to store (should be 100KB+ for optimal use).
+   */
+  async setItem(namespace: string, key: string, value: unknown): Promise<void> {
+    // Adapter handles serialization and wrapping with metadata
+    await this.#storage.setItem(namespace, key, value);
+
+    // Publish event so other controllers can react to changes
+    // Event type: StorageService:itemSet:namespace
+    // Payload: [key, value]
+    this.#messenger.publish(
+      `${SERVICE_NAME}:itemSet:${namespace}` as const,
+      key,
+      value,
+    );
+  }
+
+  /**
+   * Retrieve data from storage.
+   *
+   * Returns `unknown` since there's no schema validation.
+   * Callers should validate or cast the result to the expected type.
+   *
+   * @param namespace - Controller namespace (e.g., 'SnapController').
+   * @param key - Storage key (e.g., 'npm:@metamask/example-snap:sourceCode').
+   * @returns Parsed data or null if not found. Type is `unknown` - caller must validate.
+   */
+  async getItem(namespace: string, key: string): Promise<unknown> {
+    // Adapter handles deserialization and unwrapping
+    return await this.#storage.getItem(namespace, key);
```

Comment:
If we are to use the `itemSet` event to notify subscribers of removals, we'll need to publish here as well

Reply:
Essentially what i meant was that if you do something like this:

```
await this.messenger.call(
  'StorageService:removeItem',
  'MyController',
  'data-key',
);
```

Then we should emit `null` (or undefined?) for the event `StorageService:MyController:itemSet:data-key` so that subscribers have a way to discover it was deleted

Reply:
Yes exactly. What I meant is if there is one less element in a certain piece of data, it is still updated/stored to disk via itemSet which triggers the event. Makes sense?